File: mesp/runner.py
#
# For licensing see accompanying LICENSE file.
# Copyright (C) 2025 Apple Inc. All Rights Reserved.
#
from __future__ import annotations

from dataclasses import dataclass, asdict, field
from collections import defaultdict
from typing import Callable
from abc import ABC, abstractmethod

import mlx.core as mx
from mesp.utils import get_grad_name


@dataclass
class RunFunctionConfig:
    function_name: str
    input_names: list[str]
    output_names: list[str]
    checkpoints_to_remove: list[str] = field(default_factory=list)

    def to_dict(self) -> dict:
        return asdict(self)


def infer_checkpoints_lifetime(configs: list[RunFunctionConfig]):
    output_lifetime = {}

    all_output_names = set()
    for i, config in enumerate(configs):
        for input_name in config.input_names:
            # If a function input depends on the output from previous function
            if input_name in all_output_names:
                # The output has to be alive for this run config
                output_lifetime[input_name] = max(i, output_lifetime.get(input_name, 0))
        all_output_names.update(config.output_names)

    lifetime_to_output_names = defaultdict(list)
    for output_name, lifetime in output_lifetime.items():
        lifetime_to_output_names[lifetime].append(output_name)

    for i, config in enumerate(configs):
        if i in lifetime_to_output_names:
            config.checkpoints_to_remove.extend(lifetime_to_output_names[i])


class FunctionRunner:
    def __init__(self, function_path: str):
        self.function = mx.import_function(function_path)

    def run(self, inputs: dict[str, mx.array], params: dict[str, mx.array], config: RunFunctionConfig) -> dict[str, mx.array]:
        args = []
        for input_name in config.input_names:
            if input_name in inputs:
                args.append(inputs[input_name])
            elif input_name in params:
                args.append(params[input_name])
            else:
                raise ValueError(f"Missing input {input_name} for function {config.function_name} in {params.keys()}")
        outputs = self.function(*args)
        assert len(outputs) == len(config.output_names)
        return dict(zip(config.output_names, outputs))


@dataclass
class GradientsContext(ABC):
    pass


class BaseRunner(ABC):
    def __init__(
        self,
        function_paths: list[str],
        trainable_params_loader: Callable[[], dict[str, mx.array]],
        params_loaders: list[Callable[[], dict[str, mx.array]]],
        configs: list[RunFunctionConfig],
    ):
        self.trainable_params_loader = trainable_params_loader
        self.trainable_params = trainable_params_loader()
        self.function_runners: dict[str, FunctionRunner] = {}
        self.params_loaders: list[Callable[[], dict[str, mx.array]]] = params_loaders
        self.configs: list[RunFunctionConfig] = configs
        for path, config in zip(function_paths, configs):
            if config.function_name not in self.function_runners:
                self.function_runners[config.function_name] = FunctionRunner(path)
        self.checkpoints = {}

    def _run(self, inputs: dict[str, mx.array]):
        self.checkpoints = inputs
        for i, config in enumerate(self.configs):
            params_loader = self.params_loaders[i]
            function_runner = self.function_runners[config.function_name]
            params = params_loader() if params_loader else {}
            outputs = function_runner.run(self.checkpoints, {
                **params, **self.trainable_params
            }, config)
            self.checkpoints.update(outputs)
            for checkpoint in config.checkpoints_to_remove:
                del self.checkpoints[checkpoint]

    def get_metrics(self, metric_names: list[str]) -> dict[str, mx.array]:
        metrics = {}
        for metric_name in metric_names:
            if metric_name in self.checkpoints:
                metrics[metric_name] = self.checkpoints[metric_name]
        return metrics

    @staticmethod
    def accumulate(
        accumulated_grads: dict[str, mx.array] | None,
        grads: dict[str, mx.array]
    ) -> dict[str, mx.array]:
        if accumulated_grads is None:
            return grads

        for name in accumulated_grads:
            assert name in grads
            accumulated_grads[name] += grads[name]

        return accumulated_grads

    @abstractmethod
    def gradients(
        self, inputs: dict[str, mx.array], metric_names: list[str], context: GradientsContext
    ) -> tuple[dict[str, mx.array], dict[str, mx.array]]:
        """ Get gradients of the trainable parameters. """

    def sgd_step(self, accumulated_grads: dict[str, mx.array], learning_rate: float):
        learning_rate = mx.array(learning_rate)
        for name, grad in accumulated_grads.items():
            assert name in self.trainable_params
            self.trainable_params[name] -= learning_rate.astype(grad.dtype) * grad

    def init_adamw_state(self):
        """Initialize AdamW optimizer state (first and second moment estimates)."""
        self.adamw_m = {}  # First moment (mean of gradients)
        self.adamw_v = {}  # Second moment (variance of gradients)
        self.adamw_step_count = 0
        for name, param in self.trainable_params.items():
            self.adamw_m[name] = mx.zeros_like(param)
            self.adamw_v[name] = mx.zeros_like(param)

    def adamw_step(
        self,
        accumulated_grads: dict[str, mx.array],
        learning_rate: float,
        beta1: float = 0.9,
        beta2: float = 0.999,
        eps: float = 1e-8,
        weight_decay: float = 0.01,
    ):
        """
        AdamW optimizer step.

        Args:
            accumulated_grads: Gradients for trainable parameters
            learning_rate: Learning rate
            beta1: Exponential decay rate for first moment (default: 0.9)
            beta2: Exponential decay rate for second moment (default: 0.999)
            eps: Small constant for numerical stability (default: 1e-8)
            weight_decay: Weight decay coefficient (default: 0.01)
        """
        # Initialize state if needed
        if not hasattr(self, 'adamw_m'):
            self.init_adamw_state()

        self.adamw_step_count += 1
        t = self.adamw_step_count

        # Bias correction terms
        bc1 = 1 - beta1 ** t
        bc2 = 1 - beta2 ** t

        for name, grad in accumulated_grads.items():
            assert name in self.trainable_params
            param = self.trainable_params[name]

            # Update biased first moment estimate
            self.adamw_m[name] = beta1 * self.adamw_m[name] + (1 - beta1) * grad
            # Update biased second moment estimate
            self.adamw_v[name] = beta2 * self.adamw_v[name] + (1 - beta2) * (grad * grad)

            # Bias-corrected estimates
            m_hat = self.adamw_m[name] / bc1
            v_hat = self.adamw_v[name] / bc2

            # AdamW update: decouple weight decay from gradient update
            # weight decay applied to parameter directly
            update = m_hat / (mx.sqrt(v_hat) + eps)
            self.trainable_params[name] = (
                param * (1 - learning_rate * weight_decay) - learning_rate * update
            ).astype(param.dtype)

    def run(
        self,
        batched_inputs: list[dict[str, mx.array]],
        metric_names: list[str],
        learning_rate: float,
        num_steps: int,
        num_accumulation_steps: int,
        gradients_context: GradientsContext,
    ) -> list[dict[str, mx.array]]:
        accumulated_grads = None
        metrics_list = []
        for i in range(num_steps):
            input_idx = i % len(batched_inputs)
            grads, metrics = self.gradients(batched_inputs[input_idx], metric_names, gradients_context)
            metrics_list.append(metrics)
            # accumulate gradients
            accumulated_grads = self.accumulate(accumulated_grads, grads)
            if (i + 1) % num_accumulation_steps == 0:
                # optimize
                self.sgd_step(accumulated_grads, learning_rate)
                # reset accumulated gradients
                accumulated_grads = None

            print(f"Step: {i+1}, metrics: {metrics}")

        return metrics_list


@dataclass
class MeBPContext(GradientsContext):
    pass


class MeBPRunner(BaseRunner):
    def gradients(
        self, inputs: dict[str, mx.array], metric_names: list[str], context: MeBPContext
    ) -> tuple[dict[str, mx.array], dict[str, mx.array]]:
        self._run(inputs)
        metrics = self.get_metrics(metric_names)
        grads = {
            name: self.checkpoints[get_grad_name(name)]
            for name in self.trainable_params
        }
        self.checkpoints = {}
        return grads, metrics

File: MLXMeSP/Runner.swift
//
//  For licensing see accompanying LICENSE file.
//  Copyright (C) 2025 Apple Inc. All Rights Reserved.
//
//  Runner.swift
//  mlx-mesp
//
//  Created by Congzheng Song on 9/18/25.
//

import Foundation
import MLX

#if XTOOL_MOBILE
internal typealias MeSPImportedFunction = XToolImportedFunction
#else
internal typealias MeSPImportedFunction = ImportedFunction
#endif

/// Protocol for checkpoint storage to allow swapping between in-memory and mmap implementations
public protocol MLXArrayDictionary {
    subscript(key: String) -> MLXArray? { get set }
    mutating func merge(_ other: [String: MLXArray], uniquingKeysWith combine: (MLXArray, MLXArray) throws -> MLXArray) rethrows
    var count: Int { get }
    var isEmpty: Bool { get }
    mutating func removeAll(keepingCapacity keepCapacity: Bool)
    mutating func removeValue(forKey: String) -> MLXArray?
}

extension MLXArrayDictionary {
    // Convenience method with default parameter
    mutating func removeAll() {
        removeAll(keepingCapacity: false)
    }
}

extension Dictionary: MLXArrayDictionary where Key == String, Value == MLXArray {
    // Dictionary's built-in removeAll(keepingCapacity:) satisfies the protocol requirement
}

extension MMapCheckpointStorage: MLXArrayDictionary {
    public func removeAll(keepingCapacity _: Bool = false) { cleanup() }

    public func removeValue(forKey: String) -> MLXArray? {
        removeArray(for: forKey)
        return nil
    }
}

public struct RunFunctionConfig: Codable {
    public let functionName: String
    public let inputNames: [String]
    public let outputNames: [String]
    public var checkpointsToRemove: [String]
    
    enum CodingKeys: String, CodingKey {
        case functionName = "function_name"
        case inputNames = "input_names"
        case outputNames = "output_names"
        case checkpointsToRemove = "checkpoints_to_remove"
    }
    
    public init(functionName: String, inputNames: [String], outputNames: [String], checkpointsToRemove: [String] = []) {
        self.functionName = functionName
        self.inputNames = inputNames
        self.outputNames = outputNames
        self.checkpointsToRemove = checkpointsToRemove
    }
}

// MARK: - MeBP Runner Implementation

public typealias IterationStartHook = (Int) throws -> Void
public typealias IterationEndHook = (Int, [String: MLXArray]) throws -> Void

public class BaseRunner<Context> {
    internal var functions: [String: MeSPImportedFunction] = [:]
    internal let paramsPaths: [String?]
    internal let configs: [RunFunctionConfig]
    public var checkpoints: MLXArrayDictionary

    public let trainableParamsLoader: () throws -> [String: MLXArray]
    public var trainableParams: [String: MLXArray]
    internal let paramsLoaders: [() throws -> [String: MLXArray]]
    
    public enum StorageType {
        case inMemory
        case mmap(baseDirectory: URL? = nil)
    }

    public init(
        functionPaths: [String],
        trainableParamsLoader: @escaping () throws -> [String: MLXArray],
        paramsLoaders: [() throws -> [String: MLXArray]],
        configs: [RunFunctionConfig],
        storageType: StorageType,
    ) throws {
        self.trainableParamsLoader = trainableParamsLoader
        self.trainableParams = try trainableParamsLoader()
        self.paramsLoaders = paramsLoaders
        self.configs = configs
        self.paramsPaths = Array(repeating: nil, count: configs.count)
        
        for (path, config) in zip(functionPaths, configs) {
            if functions[config.functionName] == nil {
                functions[config.functionName] = try MeSPImportedFunction(url: URL(fileURLWithPath: path))
            }
        }
        // Initialize checkpoint storage based on type
        switch storageType {
        case .inMemory:
            checkpoints = [:]
        case .mmap(let baseDirectory):
            checkpoints = try MMapCheckpointStorage(baseDirectory: baseDirectory)
        }
    }
    
    internal func runFunction(
        _ f: MeSPImportedFunction,
        inputs: MLXArrayDictionary,
        params: MLXArrayDictionary,
        config: RunFunctionConfig,
        eval: Bool = true,
        clearCache: Bool = true
    ) throws -> [String: MLXArray] {
        var args: [MLXArray] = []

        for inputName in config.inputNames {
            if let input = inputs[inputName] {
                args.append(input)
            } else if let param = params[inputName] {
                args.append(param)
            } else {
                throw MLXError.caught("Missing input \(inputName) for function \(config.functionName)")
            }
        }

        var outputs = try f.call(args: args, kwargs: [:])

        guard outputs.count == config.outputNames.count else {
            throw MLXError.caught("Output count mismatch: expected \(config.outputNames.count), got \(outputs.count)")
        }

        // Workaround: Convert embedding output to match model dtype if needed
        if config.functionName == "embedding" {
            var expectedDtype: DType? = nil
            for inputName in config.inputNames {
                if let param = params[inputName], param.dtype == .bfloat16 || param.dtype == .float16 {
                    expectedDtype = param.dtype
                    break
                }
            }

            if let targetDtype = expectedDtype {
                outputs = outputs.map { output in
                    let elementCount = output.shape.reduce(1, *)
                    let dataSize = output.asData().data.count
                    let bytesPerElement = dataSize / elementCount
                    let needsConversion = output.dtype != targetDtype || bytesPerElement != 2

                    if needsConversion {
                        let converted = output.asType(targetDtype)
                        converted.eval()
                        return converted
                    }
                    return output
                }
            }
        }

        // Workaround: Convert transformer_block outputs to bfloat16 if needed
        if config.functionName == "transformer_block_forward" || config.functionName == "transformer_block_backward" {
            var expectedDtype: DType? = nil
            for inputName in config.inputNames {
                if let param = params[inputName], param.dtype == .bfloat16 || param.dtype == .float16 {
                    expectedDtype = param.dtype
                    break
                }
            }

            if let targetDtype = expectedDtype {
                outputs = outputs.map { output in
                    output.eval()
                    let elementCount = output.shape.reduce(1, *)
                    let dataSize = output.asData().data.count
                    let bytesPerElement = elementCount > 0 ? dataSize / elementCount : 0

                    if bytesPerElement == 4 && (targetDtype == .bfloat16 || targetDtype == .float16) {
                        let converted = output.asType(targetDtype)
                        converted.eval()
                        return converted
                    }
                    return output
                }
            }
        }

        // Eval the function outputs each time to avoid loading all unread arrays in memory
        if eval {
            outputs.forEach { $0.eval() }
        }
        // NOTE: clearCache moved to end of step() for accurate peak memory measurement
        // if clearCache { GPU.clearCache() }

        return Dictionary(uniqueKeysWithValues: zip(config.outputNames, outputs))
    }
    
    internal func step(inputs: [String: MLXArray], eval: Bool = true, debugMemory: Bool = false) async throws {
        checkpoints.removeAll()
        inputs.forEach { checkpoints[$0.key] = $0.value }

        // Memory debug helper
        func getMemoryMB() -> Double {
            return Double(GPU.activeMemory) / (1024 * 1024)
        }

        if debugMemory {
            GPU.resetPeakMemory()
            print("\n=== STEP START (checkpoints: \(checkpoints.count), mem: \(String(format: "%.0f", getMemoryMB()))MB) ===")
        }

        for (i, config) in configs.enumerated() {
            try Task.checkCancellation()
            let paramsLoader = paramsLoaders[i]
            guard let f = functions[config.functionName] else {
                throw MLXError.caught("Function runner not found for \(config.functionName)")
            }

            var params = try paramsLoader()
            params.merge(trainableParams) { (_, new) in new }

            let outputs = try runFunction(f, inputs: checkpoints, params: params, config: config, eval: eval)
            checkpoints.merge(outputs) { _, new in new }

            // Remove checkpoints as specified in config
            let removedCount = config.checkpointsToRemove.count
            for checkpoint in config.checkpointsToRemove {
                _ = checkpoints.removeValue(forKey: checkpoint)
            }

            if debugMemory {
                let phase = config.functionName.contains("backward") ? "BWD" : "FWD"
                let removal = removedCount > 0 ? "removed:\(removedCount)" : "kept"
                print("[\(String(format: "%02d", i))] \(phase) \(removal) → ckpts:\(checkpoints.count) mem:\(String(format: "%.0f", getMemoryMB()))MB")
            }
        }

        if debugMemory {
            let peakMB = Double(GPU.peakMemory) / (1024 * 1024)
            print("=== STEP END (peak: \(String(format: "%.0f", peakMB))MB) ===\n")
        }
    }
    
    internal func accumulate(accumulatedGrads: inout [String: MLXArray], grads: [String: MLXArray]) {
        for (name, grad) in grads {
            if let currentGrad = accumulatedGrads[name] {
                accumulatedGrads[name] = currentGrad + grad
            } else {
                accumulatedGrads[name] = grad
            }
        }
    }
    
    internal func sgdStep(accumulatedGrads: [String: MLXArray], learningRate: Float, numAccumulationSteps: Int) {
        let lr = MLXArray(learningRate / Float(numAccumulationSteps))
        
        for (name, grad) in accumulatedGrads {
            if let param = trainableParams[name] {
                trainableParams[name] = param - lr.asType(grad.dtype) * grad
            }
        }
    }
    
    internal func getGradName(_ name: String) -> String {
        return "\(name).grad"
    }
    
    internal func getMetrics(metricsName: [String]) -> [String: MLXArray] {
        var metrics = [String: MLXArray]()
        for name in metricsName {
            if let metric = checkpoints[name] {
                metrics[name] = metric
            }
        }
        return metrics
    }
    
    public func gradients(
        inputs: [String: MLXArray],
        metricNames: [String],
        context: Context,
    ) async throws -> (grads: [String: MLXArray], metrics: [String: MLXArray]) {
        throw MLXError.caught("Not implemented.")
    }

    public func run(
        batchedInputs: [[String: MLXArray]],
        metricNames: [String],
        learningRate: Float,
        numSteps: Int,
        numAccumulationSteps: Int,
        gradientsContext: Context,
        onIterationStart: IterationStartHook? = nil,
        onIterationEnd: IterationEndHook? = nil,
        verbose: Bool = false,
    ) async throws -> [[String: MLXArray]] {
        var accumulatedGrads: [String: MLXArray] = [:]
        var metricsList: [[String: MLXArray]] = []
        
        for i in 0..<numSteps {
            // Check for task cancellation and throw to properly exit
            try Task.checkCancellation()
            try onIterationStart?(i)
            let inputIdx = i % batchedInputs.count
            let (grads, metrics) = try await gradients(inputs: batchedInputs[inputIdx], metricNames: metricNames, context: gradientsContext)
            metricsList.append(metrics)
            // Accumulate gradients
            accumulate(accumulatedGrads: &accumulatedGrads, grads: grads)
            if (i + 1) % numAccumulationSteps == 0 {
                // Optimize
                sgdStep(accumulatedGrads: accumulatedGrads, learningRate: learningRate, numAccumulationSteps: numAccumulationSteps)
                // Reset accumulated gradients
                accumulatedGrads = [:]
            }
            try onIterationEnd?(i, metrics)
        }
        return metricsList
    }

}

public struct MeBPContext {
    public var debugMemory: Bool = false

    public init(debugMemory: Bool = false) {
        self.debugMemory = debugMemory
    }
}

public class MeBPRunner: BaseRunner<MeBPContext> {
    override public func gradients(
        inputs: [String: MLXArray],
        metricNames: [String],
        context: MeBPContext
    ) async throws -> (grads: [String: MLXArray], metrics: [String: MLXArray]){
        try await step(inputs: inputs, debugMemory: context.debugMemory)
        let metrics = getMetrics(metricsName: metricNames)
        var grads: [String: MLXArray] = [:]
        for name in trainableParams.keys {
            if let grad = checkpoints[getGradName(name)] {
                grads[name] = grad
            }
        }
        checkpoints.removeAll()
        return (grads: grads, metrics: metrics)
    }

    public func run(
        batchedInputs: [[String: MLXArray]],
        metricNames: [String],
        learningRate: Float,
        numSteps: Int,
        numAccumulationSteps: Int,
        onIterationStart: IterationStartHook? = nil,
        onIterationEnd: IterationEndHook? = nil,
        verbose: Bool = false,
        debugMemory: Bool = true,
    ) async throws -> [[String: MLXArray]] {
        try await run(
            batchedInputs: batchedInputs,
            metricNames: metricNames,
            learningRate: learningRate,
            numSteps: numSteps,
            numAccumulationSteps: numAccumulationSteps,
            gradientsContext: MeBPContext(debugMemory: debugMemory),
            onIterationStart: onIterationStart,
            onIterationEnd: onIterationEnd,
            verbose: verbose
        )
    }
}

/// Training metadata loaded from `training_config.json` produced by the
/// Python export pipeline. Optional fields keep older exports loadable.
public struct TrainingConfig: Codable {
    public let numLayers: Int
    public let mode: String
    public let swiftStructuredBackward: Bool
    public let loraRank: Int
    public let seqLength: Int
    public let batchSize: Int

    enum CodingKeys: String, CodingKey {
        case numLayers = "num_layers"
        case mode
        case swiftStructuredBackward = "swift_structured_backward"
        case loraRank = "lora_rank"
        case seqLength = "seq_length"
        case batchSize = "batch_size"
    }

    public init(from decoder: Decoder) throws {
        let c = try decoder.container(keyedBy: CodingKeys.self)
        self.numLayers = try c.decode(Int.self, forKey: .numLayers)
        self.mode = (try? c.decode(String.self, forKey: .mode)) ?? "mebp"
        self.swiftStructuredBackward =
            (try? c.decode(Bool.self, forKey: .swiftStructuredBackward)) ?? false
        self.loraRank = (try? c.decode(Int.self, forKey: .loraRank)) ?? 8
        self.seqLength = (try? c.decode(Int.self, forKey: .seqLength)) ?? 256
        self.batchSize = (try? c.decode(Int.self, forKey: .batchSize)) ?? 1
    }
}
